course6/string_ariel/str.h

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <iostream>
+#include <cstring>
+
+class Str
+{
+public:
+    Str()
+    {
+        start = mid = end = nullptr;
+    }
+
+    Str(const char* s)
+    {
+        int size = std::strlen(s);
+        start = new char[size];
+        end = start;
+
+        while (*s != '\0')
+        {
+            *end++ = *s++;
+        }
+
+        mid = end;
+    }
+
+    Str(const Str& str)
+    {
+        int size = str.Size();
+        start = new char[size];
+        end = start;
+
+        for (int i = 0; i < size; i++)
+        {
+            *end++ = str[i];
+        }
+
+        mid = end;
+    }
+
+    Str(Str&& str) :start(str.start), mid(str.mid), end(str.end)
+    {
+        str.start = str.mid = str.end = nullptr;
+    }
+
+    ~Str()
+    {
+        Dispose();
+    }
+
+    char operator[](int i) const
+    {
+        return *(start + i);
+    }
+
+    char& operator[](int i)
+    {
+        return *(start + i);
+    }
+
+    Str& operator = (const Str& str)
+    {
+        if (str.Size() > Capacity())
+        {
+            Str temp(str);
+            Swap(*this, temp);
+        }
+        else
+        {
+            mid = start + str.Size();
+            Copy(str);
+        }
+
+        return *this;
+    }
+
+    Str& operator = (const char* s)
+    {
+        int size = std::strlen(s);
+        if (size > Capacity())
+        {
+            Str temp(s);
+            Swap(*this, temp);
+        }
+        else
+        {
+            mid = start + size;
+            Copy(s);
+        }
+
+        return *this;
+    }
+
+    Str operator + (const Str& rh)
+    {
+        return Str(*this, rh);
+    }
+
+    Str& operator += (const Str& rh)
+    {
+        if(rh.Size() + Size() <= Capacity())
+        {
+            for(int i = 0; i < rh.Size(); i++)
+            {
+                *mid++ = rh[i];
+            }
+        }
+        else
+        {
+            Str temp = *this + rh;
+            Swap(*this, temp);
+        }
+
+        return *this;
+    }
```

Comment:
and this is almost identical to the private concatenating constructor

Reply:
I am not completely sold, the concatenating constructor suits well inside the else statement (which is what the + operator does) but inside the if statement I am no really interested in allocating memory, that’s why I didnt use the constructor

Reply:
My point is: I see the same code twice (or even thrice). This shouldn't be. There is many ways of avoiding duplication.